### ISSUE TITLE
[GNU/Linux] [Possible solution] Gtk emoji picker doesn't work on AppImage but works on dev mode

### ISSUE THREAD

**Issue by meow464** [beeware/briefcase#481]
Following the hello world example from: https://docs.beeware.org/en/latest/tutorial/tutorial-1.html
If you right click the text input widget and select "Insert Emoji", the AppImage version crashes. In dev mode you get gtk emoji picker, as expected.
The crash log on the terminal is the following (I had to translate some parts to English):
```
[GTK+] Not implemented: Widget.set_hidden()
[GTK+] Not implemented: Widget.set_hidden()
[GTK+] Not implemented: TextInput.set_font()
[GTK+] Not implemented: TextInput.set_font()
[GTK+] Not implemented: Widget.set_hidden()
[GTK+] Not implemented: TextInput.set_font()
[GTK+] Not implemented: TextInput.set_alignment()
[GTK+] Not implemented: TextInput.set_font()
[GTK+] Not implemented: TextInput.set_font()
[GTK+] Not implemented: Widget.set_hidden()
[GTK+] Not implemented: Widget.set_hidden()
[GTK+] Not implemented: Button.set_background_color()

(__main__.py:13824): Gtk-WARNING **: 12:19:22.975: Could not load a pixbuf from icon theme.                                                                                       
This may indicate that pixbuf loaders or the mime database could not be found.
**
Gtk:ERROR:../../../../gtk/gtkiconhelper.c:494:ensure_surface_for_gicon: assertion failed (error == NULL): Failed to load /home/user/.icons/elementary-xfce/status/16/image-missing.svg: [TRANSLATED] Can't load the image loading module [\TRANSLATED]: /usr/lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so: /usr/lib/x86_64-linux-gnu/librsvg-2.so.2: undefined symbol: cairo_tag_end (gdk-pixbuf-error-quark, 5)
/tmp/.mount_Hello_4a1FKE/AppRun, linha 7: 13824 Abortado                $APPDIR/usr/bin/python3 -s -m helloworld   
```

Debian 10 (buster) amd64;
Python3.7;
Everything beeware was installed with pip on a virtual env.

**Comment by ceo-of-programming** [comment]
Found possible cause (5 minutes after opening the issue). It works if I do:
```
cd /path/to/hello_world
hello_world.AppImage --AppImage-extract
rm squashfs-root/usr/lib/libcairo.so.2
squashfs-root/AppRun
```
Reason: The docker container gets everything python with pip (the newest possible) but libcairo comes from the ubuntu repos. "cairo_tag_end" was added in 2018 with the release of libcairo 1.16, while the version on the ubuntu repos is 1.14.
Cairo release notes mention the addition of "cairo_tag_end": https://www.cairographics.org/news/cairo-1.16.0/

Solutions:
1. Install an older version of whichever python package pulls this string;
2. Bump the docker OS to Debian buster (released in 07/2019). Ubuntu 19.10 is not LTS.
3. Try to install the newest libcairo on 16.04 and hope it doesn't destroy the system (it isn't on backports).

**Comment by freakboy3742** [comment]
Thanks for that report... it's definitely interesting that deleting libcairo like that resolves the problem. We've had similar reports recently (#470) which points at a problem with the way we're driving AppImage (or a problem with AppImage itself...)

AppImage is *supposed* to isolate the binaries for the app. The app itself is compiled inside a Docker container using Ubuntu 16.04, and other than fundamental libraries like libc, all library dependencies *should* be included in the AppImage, and be delocalized so they don't contain references to libraries outside the AppImage bundle.

So - the fact that removing libcairo resolves this problem suggests that there is something "leaking" - something isn't being correctly isolated from the cairo libraries.

Your proposed solutions would almost certainly *work* - but they would also remove the purpose of using AppImage in the first place. The point of AppImage is to build on the "oldest libc you want to support", and the AppImage will work on any Linux newer than that. This includes running on Fedora and other non-Ubuntu Linuxen. This is something that *has* worked historically; although I'm guessing it probably doesn't work at the moment for very similar reasons to the one you've highlighted.

**Comment by ceo-of-programming** [comment]
> Thanks for that report... it's definitely interesting that deleting libcairo like that resolves the problem. 

From my understanding, an AppImage is just an squashfs image with libs, it gets mounted and the ld_path is set to whatever is inside it. If something is missing, the system looks for it on the default ld_path.

>We've had similar reports recently (#470) which points at a problem with the way we're driving AppImage (or a problem with AppImage itself...)

The underlying problem is that AppImage only copies stuff from the build system inside an squashfs image, it assumes that said stuff is in fact working.

>AppImage is supposed to isolate the binaries for the app. The app itself is compiled inside a Docker container using Ubuntu 16.04, and other than fundamental libraries like libc, all library dependencies should be included in the AppImage, and be delocalized so they don't contain references to libraries outside the AppImage bundle.

I will test my guess that the hello world example, with the newest python libs, doesn't actually work on 16.04. If that is the case, we can only try to backport libcairo to 16.04, use an older version of whichever python package needs libcairo or use a newer system.

>Your proposed solutions would almost certainly work - but they would also remove the purpose of using AppImage in the first place. The point of AppImage is to build on the "oldest libc you want to support", and the AppImage will work on any Linux newer than that.

The libcairo we are packaging does work with 16.04's libc, I think the problem is that the python part doesn't work with with 16.04's libcairo. That is because the python part is not actually part of 16.04, it was installed with pip and is much more recent.
Incidentally, this problem we are hitting is one argument in favor of only installing things through a distro's package manager. Everything on the package manager is (or should) be tested against the other relevant components on the system.

**Comment by freakboy3742** [comment]
Your understanding about AppImage ld_path is correct - except that the build process includes *discovering* all the libraries that are needed, and copying them (and modifying them) so the app is self contained. The app *shouldn't* ever be looking outside the image itself, except for libraries like libc.

As I said - there should *not* be a need to back port *anything*. If the app runs on an actual 16.04 machine, it *should* run on *any* linux with a same-version-or-newer libc - and in the recent past, this *was* true. However, we've also made some recent changes to the way we invoke the AppImage build that might be affecting things. Running a build with Briefcase 0.3.2 should have the "old" packaging behavior, which was known to work.

**Comment by ceo-of-programming** [comment]
>it should run on any linux with a same-version-or-newer libc - and in the recent past, this was true

Well, it runs, it is just the emoji picker that crashes it.

**Comment by freakboy3742** [comment]
Sure - but in the past, it ran and *didn't crash*.

**Comment by ceo-of-programming** [comment]
Many problems, very issues. I hate computers.

Well, it does work on Xubuntu 16.04, but the gtk version shipped with it does not have an emoji picker. 
The last verison of toga_gtk that does not explicitly state libcairo>=1.17 as a requirement is 0.3.0.dev8. It still works in dev mode but now I can't package it, because:
1: briefcase 0.30.0.dev8 doesn't seem to use docker;
2: it needs libcrypt0.so.1.0.0 but that is only available on ubuntu 16.04 (shipping an outdated crypto library in the appimage is not great either).

**Comment by ceo-of-programming** [comment]
I don't know what to do now. The only solution I can see is changing the OS on the container. 
In the meantime is there a way for me to specify my own docker file or my my own briefcase "build file"? I'm working on an instant messaging program and emojis are important.

**Comment by freakboy3742** [comment]
Ah - ok, so you're in a slightly different situation, where you *need* the feature that wasn't available in the 16.04 build of GTK/Cairo.

In that case - yes, you'll need to use a customised Docker file.

If the package you need is available on the Ubuntu 16.04 repository, then you can add the dependency to the `system_requires` list, and it will be installed and used as part of the AppImage build. You could also modify the base OS image of the Docker container (accepting that you'll lose potential compatibility with older operating systems).

However, I'm guessing this isn't an option in this case; in which case, you will need a custom Dockerfile. The process isn't *that* complicated:
* Fork the beeware/briefcase-linux-appimage-template project
* Modify the `Dockerfile` to add whatever extra packages/repositories are needed to install the version of libcairo you need
* Modify your `pyproject.toml` to include the line:

    template = "http://github.com/yourname/briefcase-linux-appimage-template"

  in the `[...linux]` section of your app config. This will direct your app to use your customised version of the app template.

For the purposes of rapid iteration, you can also set `template` to point at a local checkout of the template. 

Alternatively, you can `briefcase create`, then manually modify the Dockerfile that is generated. This is much easier to set up, but the risk is that you'll accidentally delete/overwrite your configuration.

**Comment by damascene** [comment]
I'm having similar issue with slightly different error message, my app shows an svg image.
```
Traceback (most recent call last):
  File "/tmp/.mount_Hello_8ogeF7/usr/app_packages/toga_gtk/widgets/button.py", line 39, in gtk_on_press
    self.interface.on_press(self.interface)
  File "/tmp/.mount_Hello_8ogeF7/usr/app_packages/toga/handlers.py", line 55, in _handler
    result = handler(interface, *args, **kwargs)
  File "/tmp/.mount_Hello_8ogeF7/usr/app/helloworld/app.py", line 150, in add_image
    self.image_view = toga.ImageView(image, style=Pack(padding=10, width=300, height=300))
  File "/tmp/.mount_Hello_8ogeF7/usr/app_packages/toga/widgets/imageview.py", line 26, in __init__
    self.image = image
  File "/tmp/.mount_Hello_8ogeF7/usr/app_packages/toga/widgets/imageview.py", line 42, in image
    self._image.bind(self.factory)
  File "/tmp/.mount_Hello_8ogeF7/usr/app_packages/toga/images.py", line 36, in bind
    self._impl = factory.Image(interface=self, path=full_path)
  File "/tmp/.mount_Hello_8ogeF7/usr/app_packages/toga_gtk/images.py", line 13, in __init__
    self.native = GdkPixbuf.Pixbuf.new_from_file(str(path))
gi.repository.GLib.Error: gdk-pixbuf-error-quark: Unable to load image-loading module: /usr/lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so: /lib/x86_64-linux-gnu/librsvg-2.so.2: undefined symbol: cairo_font_options_get_variations (5)
```

**Comment by damascene** [comment]
I confirm solution above worked for me:
https://github.com/beeware/briefcase/issues/481#issuecomment-680247541

**Comment by dynobo** [comment]
I'm having the same issue as @damascene, but I struggle with applying the solution.

I would like to try solution 2 from [#481 (comment) ](https://github.com/beeware/briefcase/issues/481#issuecomment-680247541):
>  Bump the docker OS to Debian buster (released in 07/2019). Ubuntu 19.10 is not LTS.

But how can I adjust the docker OS for the briefcase workflow? I can modify the image in `./linux/appimage/myapp/Dockerfile`, but `briefcase create` will just overwrite it, and `briefcase build` will not apply the change.

Any help for me (<= briefcase-newbie) is appreciated!

**Comment by freakboy3742** [comment]
@dynobo As of late last year (around Nov 21 2020), the default Briefcase image should be based on Ubuntu 18.04, so this issue *should* be resolved... are you saying you're still seeing the problem?

As for how to modify the base image version: one of the options to Briefcase is `template` - that will specify the app template that is used. By default, this is `https://github.com/beeware/briefcase-linux-appimage-template` for Linux; however, you can point Briefcase at any GitHub repository or local file path. If you fork the default template and make whatever changes you want - or clone the default and make changes locally - you can point your project to use your own version of the app template.

**Comment by dynobo** [comment]
@freakboy3742 Thanks for the info about how to modify the base image, currently I'm using [the workaround proposed earlier](https://github.com/beeware/briefcase/issues/481#issuecomment-680247541): I just extract the AppImage,  remove the `libcairo.so.2` and pack it again using appimagetool. That works fine for me, still it would be even better, if it would be fixed at the source :-)

**Comment by ceo-of-programming** [comment]
Ubuntu 18.04 has libcairo 1.15, not 1.16. It's not on backports either. 

Debian Buster (stable) on the other hand has libcairo 1.16. It was released 15 months after ubuntu 18.04, a bit younger, not terrible but not great either.

**Comment by freakboy3742** [comment]
I'm going to close this issue to consolidate the general problem of "Linux AppImage deployment has problems" under a single issue (#718).
